design/notifications.md

```diff
@@ -0,0 +1,405 @@
+
+##### `POST /delete-all`
+
+This endpoint marks all notifications for a user as having been deleted. See [the notification agent README][8] for
+details.
+
+#### API Version 2
+
+##### `GET /messages`
+
+This endpoint lists messages that have been stored in the database for a user. This endpoint will remain largely
+unchanged from its counterpart in version 1 of the API. But a few changes will be made. First the camelCased query
+parameter names will be replaced with kebab-cased names instead. Second, q new Boolean query parameter,
+`notification-order`, will be added that will cause the endpoint to apply the `limit` parameter to the query in
+descending order, but return the messages in ascending order. Setting this query parameter to `true` replaces the
+current `/last-ten-messages` endpoint with a more flexible and RESTful alternative. Third, another new Boolean
+parameter, `count-only`, will be added that instructs the endpoint to return just the message counts ratehr than the
+actual messages. Setting this query parameter to `true` replaces the functionality of the `count-messages` from version
+1 of the API. The response body will take the following form:
+
+```
+{
+    "total": 32,
+    "messages": []
+}
+```
+
+The `total` element will contain the total number of messages that match the request. This element will always be
+present in the response body. The `messages` element will contain the actual notification messages. It will not be
+present if the `count-only` query parameter is present and set to `true`. The format of each message in the list of
+messages will be the same as in API version 1, except that camelCased elements in the top level of the response will be
+replaced by snake_cased elements instead.
+
+Available query parameters:
+
+| Query Parameter    | Description                                                                                  |
+| ------------------ | -------------------------------------------------------------------------------------------- |
+| user               | The username of the user to retrieve notification messages for (required).                   |
+| limit              | The maximum number of messages to return in the response body (default: all messages).       |
+| offset             | The offset of the first message to return in the response body (default: 0).                 |
+| seen               | If `true` messages that have been seen will be included in the response (default: false).    |
+| sort-field         | The field to sort messages by (default: timestamp).                                          |
+| sort-dir           | Allows the response body to sorted in ascending or descending order (default: descending).   |
+| message-type       | If specified, only messages of the given type will be included in the response.              |
+| notification-order | If `true` messages will be sorted in ascending order after the limit has been applied.       |
```

Comment:
This one is a little tricky because it involves the limit field as well. The idea is that this will cause the endpoint to return the ten most recent notifications in ascending order by timestamp. Without this parameter, if you sort by timestamp in ascending order with an offset of 0 and a limit of 10 then you get the ten oldest messages. If you sort by timestamp in descending order with an offset of 0 and a limit of 10 then you get the ten most recent messages, but they're not in an order that's convenient for the UI to work with.

There are workarounds of course. When the feature was initially requested, I suggested two possible alternatives. It would be possible to obtain the total count then calculate the correct offset to use when sorting in ascending order. This requires multiple API calls, though, which can be sluggish. It would also be possible to sort in descending order and then sort the notifications in ascending order again on the client side. The fear at the time this was initially requested was this would consume too many resources on the client side.

Reply:
I'm not sure if it lends any additional clarity, but I feel obligated to ask whenever we're updating an API with limits and offsets: can we use [keyset pagination](https://use-the-index-luke.com/no-offset) here? It produces less of an immediately sensible "page", but it's more performant on the database end and I wonder if it might make this clearer too (and possibly more featureful). To do that here:

* right now: limit/offset/field/direction for determining the result set, then a separate field that can change direction from desc to asc for the returned value only
* keyset setup: a limit plus a set of key, predicate, value tuples (e.g. timestamp, greater than, 2020-01-01T19:54:00-07:00) which are jointly applied to determine the result set, plus a separate set of field+direction tuples for the ordering of the returned values

An intermediate option, but admittedly one I think could be more confusing, would be to keep using limit/offset/field/direction pagination, but allow fully specifying the post-fetch re-sort, rather than just allowing the one flip from desc to asc. I'm not sure this would be worth it unless we need it, but it would be more generalizable.

Anyway, mostly musing, but given the DB performance aspects (specifically: not being slower the deeper you are in the pagination) I feel like I should always at least ask about keyset pagination as an option.

Reply:
Sounds promising. Thanks for the suggestion; I'll look into it.

Reply:
> It would also be possible to sort in descending order and then sort the notifications in ascending order again on the client side.

This was my immediate thought as well. It seems like an unnecessary burden on backend devs, and unnecessary complexity in the API, to add an extra endpoint (or extra query param and extra logic) when the client can just use `sort-field` and `sort-dir` with a `limit` of 10, and simply reverse the display order in the UI if necessary*. If any device is using too many resources reversing an array of 10 items, then I seriously doubt it could even display our UI in the first place 😆 

*I don't even know why this is necessary. It seems when we show the last 10 notifications in the current UI, the newest is displayed first, and this already the result of a descending sort for the timestamp 😕

**edit: I had to check... so the API is doing an unnecessary reverse, only to be reversed again by the UI 🤦  🤦‍♀️  🤦‍♂️ 
https://github.com/cyverse-de/ui/blob/4fba87fef5522dca786ca826ce8a417b396f9ab3/react-components/src/desktop/view/Notifications.js#L219-L230

The madness can end with API v2! 😵 So let's drop `notification-order` query param support, although @ianmcorvidae's idea is still pretty interesting and maybe worth looking into.

Reply:
Cool. I honestly don't remember who requested it the feature. I'm definitely happy to nuke it, though.

Reply:
I do have a couple of questions about keyset pagination.

First, it doesn't work well with options to link directly to a specific page. I know that we'd talked about supporting that a few times. Were we still planning to support it? If so, would that be something we could reconsider, at least in the case of notifications?

Second, I can see how it would work for columns with a uniqueness constraint, but it doesn't seem like it would work very well for sorting on columns that are not unique. We sort on timestamp by default, which isn't guaranteed to be unique, but with a precise enough timestamp, we can probably make it work reasonably well. We also support sorting based on notification ID, which should work fine. The one field that concerns me is the notification subject. This isn't guaranteed to be unique, and is quite likely to have a lot of duplicate values. On the other hand, I don't really see a good reason to provide the ability to sort by subject. Does anyone see a reason we can't omit the subject from the list of sortable columns?

Reply:
on a.) one thing you can do with keyset pagination, at least, is link to a "page" that's based on the keyset. This is seen in the wild some, but it usually means a link with something like `?before=20200204T15:00:00-07:00` (or any other specification of the keyset), rather than `?page=n`. It's a little less precise in some sense, but still always possible to link to a fairly specific point in the pagination. Conveniently, it should also remain valid & showing nearly the same data, so long as nothing new gets inserted that would appear in the same section.

for b.) the standard way to do this, as I understand, is to always append at least one unique column to the end of the keyset if there isn't one already (ids are good where we can't rely on timestamps being unique). That way there's always a completely definite sort order. I do think it would be fine to remove the ability to sort by subject as well, though.

Reply:
That helps. Thanks for the additional information.

Reply:
I didn't think about this very carefully last night, so I wanted to revisit it again today. It still seems to me that keyset pagination won't work very well with linking to a specific page. Unless the client side code has paged through the results already so that it knows which timestamps correspond to page boundaries, assuming the notifications are sorted by timestamp.

I don't think this is necessarily a deal-breaker for keyset pagination, but I do think that it will require us to avoid providing links to specific page numbers. (This may have been what you meant yesterday, but I wanted to make sure that it was clear that I was arguing for a potential UI change rather than arguing against keyset pagination.) Instead of linking to specific page numbers, we'd probably want to have a widget in the UI that will allow the user to view messages older than or newer than a specific timestamp. It seems like that should provide a slightly better user experience anyway.

I read about an extension to keyset pagination that sounds promising. Rather than having the caller specify the offset to begin or end at using `before` or `after` query parameters, the UI can send `before_id` or `after_id` query parameters instead. This will give us a definite sort order even for columns that may potentially have a significant number of duplicate values. The blog post that I read referred to these as "low-cardinality columns." I think that this will provide a reasonable and reliable way for us to page through the results, even if it doesn't allow us to skip pages very easily.

The existing API also allows sorting by the subject line and notification UUID. I really don't see a good reason to be able to sort by these fields, so I'd like to propose these changes:

Reply:
I just had a thought that will at least partially take care of the problem with the edge case mentioned above. If we support both `after` and `before` query parameters and `after_id` and `before_id` query parameters then the initial query that is sent down by the UI when notifications are listed can include the current timestamp if notifications are sorted in descending order. This doesn't completely prevent a return to the first page from loading additional results, but with a precise enough timestamp, we can make it very unlikely.

Reply:
Ah, yes, I'd misunderstood the page thing. You're right, it's not really possible to link to "page 4" since the boundaries aren't divided like that any more. I think it's usually replaced with the forward/back (as expected) plus (for timestamps) stuff like "next day", "next week", or what-have-you. UI voices definitely needed for if that's good enough.

I like the idea of having all 4 of `after`, `before`, `after_id`, and `before_id`, and using the two afters or two befores jointly to decide the page. I didn't put it very well, but that's exactly the kind of thing that I meant with "appends a unique value to the end of the keyset" for a fixed ordering.

As far as paging backwards, I'd assume each page/view would only send things based on its own values. So for forward, it'd send `after`/`after_id`, and for backwards it'd send `before`/`before_id`. I'd expect everything else to be relative stuff like back to the first page (which I'd say should always show the most recent, even if that's changed), back an hour/day/etc. For the sake of nicety, maybe we'd want to pass along some notion of min/max values to the UI (so it can show the interface accordingly), and for the first page I'd imagine we'd always show a full set of results even if the `before`/`before_id` would only produce a partial page. [if I'm rambling again, ignore me, heh]

admittedly, calling this "pagination" is itself slightly misleading since there aren't really "pages" so much as slices of the data

Anyway! I think this sounds pretty good and we're converging toward something that'll be effective one way or the other, and some more UI voices would be good.

Reply:
Ah, yes, changing the sort order and switching to `before` and `before_id` would work well. The sort order would be reversed, but that shouldn't be a big deal. Ever since the request for the `last-ten-messages` endpoint, I've had a habit of assuming that no sorting would be done on the client side. I guess I haven't quite broken that habit yet. 😖

Reply:
I think the `before`/`after` keysets sound good for a feed like notifications. For something more stable, like data store listings, then specific pages make more sense, but pages might actually be less convenient for a live feed like notifications.

For example, if someone is browsing pages in the notification view, and between page-changes 10 new notifications come in, then once they browse to the next page, they'll be looking at 10 notifications they were just browsing.

With the `before`/`after` keysets, I don't think this wouldn't happen.

This is also similar to browsing https://old.reddit.com feeds, except I think they use post IDs in their `before`/`after` links (which can cause issues if the post with that ID gets removed before you have a chance to click on it).

I also don't think we need to sort on subject (the current UI doesn't provide this option), but I think being able to search on subjects would be a helpful substitute. For example, if I remembered sharing an app with a user, but I don't remember which app I shared, it would be nice to search for that user in the subject (and filtering on Apps notifications).

Reply:
With all that said, I'm still not sure how difficult this would be to implement in the UI. I don't think it would be too bad, although I think we would have build something custom since I don't think the MUI pagination tool would support this (but maybe it's just a matter of a couple of `back` / `next` links instead).

Reply:
Cool, I think I'll probably go with this proposal and add an option for filtering based on the subject rather than for sorting based on the subject.